### PR TITLE
 Add parameter statistic tracking to EventRecorder

### PR DESCRIPTION
**Pull request by jetuk**
- Adds the ability for Event objects to hold a simple list of values from
   a Parameter for the duration of the event.
   - Currently only supports Parameters and not other components or nodes.
 - New EventStatisticRecorder to compute arbitrary aggregation of the
   tracked values and return via .values() and .aggregated_value()
   methods.
 - Test of the new object and functionality.

There's also another commit tacked on that adds a non-finite value sense check to TablesArrayParameter

### CODE REVIEW

File: pywr/recorders/events.py [hunk 1]
```diff
@@ -52,11 +53,16 @@ def __init__(self, model, threshold, minimum_event_length=1, **kwargs):
...
+        # TODO make this more generic to track components or  nodes (e.g. storage volume)
+        self.tracked_parameter = tracked_parameter
+        if self.tracked_parameter is not None:
+            self.tracked_parameter.parents.add(self)
 
     def setup(self):
         pass
 
     def reset(self):
+        print('reset')
```
Comment (snorfalorpagus) [inline pywr/recorders/events.py]: Remove this.

File: pywr/recorders/events.py [hunk 2]
```diff
@@ -197,3 +211,70 @@ def finish(self):
         # ... and update the internal values
         for index, row in grouped.iterrows():
             self._values[index] = row['duration']
+
+
+class EventStatisticRecorder(Recorder):
+    """ Recorder for the duration of events found by an EventRecorder
+
+    This Recorder uses the results of an EventRecorder to calculate aggregated statistics
+    of those events in each scenario. This requires the EventRecorder to be given a tracked_parameter
```
Comment (snorfalorpagus) [inline pywr/recorders/events.py]: "tracked_parameter" should be in backticks

File: pywr/recorders/events.py [hunk 3]
```diff
@@ -197,3 +211,70 @@ def finish(self):
...
+    in order to save an array of values during each event. This recorder uses `event_agg_func` to aggregate
+    those saved values in each event before applying `recorder_agg_func` to those values in each scenario.
+    Aggregation by scenario is done via the pandas.DataFrame.groupby() method.
+
+    Any scenario which has no events will contain a NaN value regardless of the aggregation function defined.
+
+    Parameters
+    ----------
```
Comment (snorfalorpagus) [inline pywr/recorders/events.py]: model

File: pywr/recorders/events.py [hunk 4]
```diff
@@ -197,3 +211,70 @@ def finish(self):
...
+        self._values = np.empty(len(self.model.scenarios.combinations))
+
+        if self.event_recorder.tracked_parameter is None:
+            raise ValueError('To calculate event statistics requires the parent `EventRecorder` to have a `tracked_parameter`.')
+
+    def reset(self):
+        self._values[...] = np.nan
+
+    def after(self):
+        pass
```
Comment (snorfalorpagus) [inline pywr/recorders/events.py]: Remove this method.

File: pywr/recorders/events.py [hunk 5]
```diff
@@ -197,3 +211,70 @@ def finish(self):
...
+
+    def values(self):
+        return self._values
+
+    def finish(self):
+        """ Compute the aggregated value in each scenario based on the parent `EventRecorder` events """
+        events = self.event_recorder.events
+        # Return NaN if no events found
+        if len(events) == 0:
+            self._values[...] = np.nan
```
Comment (snorfalorpagus) [inline pywr/recorders/events.py]: This line shouldn't be required, as it's already set to `np.nan` in `reset`.

**Comment by snorfalorpagus** [comment]
This looks good overall. Some of the `Event*` recorders are missing documentation for the `agg_` keyword arguments. Even if they're implemented by the parent class I think we need to include them as the documentation will be read in isolation.

**Comment by jetuk** [comment]
Is this OK now?

**Pull request merged**